warehouse/internal/snapshots/staging_schema_test.go

```diff
@@ -0,0 +1,252 @@
+	m.insertCalls.Add(1)
+	return m.insertFunc(ctx, sourceID, destinationID, workspaceID, schemaBytes)
+}
+
+func (m *mockStagingFileSchemaDBRepo) GetLatest(ctx context.Context, sourceID, destinationID string) (*model.StagingFileSchemaSnapshot, error) {
+	m.getLatestCalls.Add(1)
+	return m.getLatestFunc(ctx, sourceID, destinationID)
+}
+
+type fixedExpiryStrategy struct {
+	expired bool
+}
+
+func (f *fixedExpiryStrategy) IsExpired(*model.StagingFileSchemaSnapshot) bool {
+	return f.expired
+}
+
+func TestStagingFileSchema(t *testing.T) {
+	schema := json.RawMessage(`{"foo":"bar"}`)
+	snapshot := &model.StagingFileSchemaSnapshot{
+		ID:            uuid.New(),
+		Schema:        schema,
+		SourceID:      "279L3gEKqwruBoKGsXZtSVX7vIy",
+		DestinationID: "27CHciD6leAhurSyFAeN4dp14qZ",
+		WorkspaceID:   "279L3V7FSpx43LaNJ0nIs9KRaNC",
+		CreatedAt:     time.Now(),
+	}
+
+	t.Run("cache hit", func(t *testing.T) {
+		db := &mockStagingFileSchemaDBRepo{
+			getLatestFunc: func(ctx context.Context, sourceID, destinationID string) (*model.StagingFileSchemaSnapshot, error) {
+				return nil, errors.New("should not be called")
+			},
+			insertFunc: func(ctx context.Context, sourceID, destinationID, workspaceID string, schemaBytes json.RawMessage) (uuid.UUID, error) {
+				return uuid.Nil, errors.New("should not be called")
+			},
+		}
+		cache := NewStagingFileSchema(config.New(), db, &fixedExpiryStrategy{expired: false})
+		cache.cache.Put(cacheKey("279L3gEKqwruBoKGsXZtSVX7vIy", "27CHciD6leAhurSyFAeN4dp14qZ"), snapshot, cache.cacheRefreshTTL())
+
+		got, err := cache.GetOrCreate(context.Background(), "279L3gEKqwruBoKGsXZtSVX7vIy", "27CHciD6leAhurSyFAeN4dp14qZ", "279L3V7FSpx43LaNJ0nIs9KRaNC", schema)
+		require.NoError(t, err)
+		require.Equal(t, snapshot, got, "expected cache hit to return cached snapshot")
+		require.Zero(t, db.getLatestCalls.Load(), "expected no DB calls")
+		require.Zero(t, db.insertCalls.Load(), "expected no DB insert calls")
+	})
+
+	t.Run("cache expired", func(t *testing.T) {
+		fresh := &model.StagingFileSchemaSnapshot{
+			ID:            uuid.New(),
+			Schema:        schema,
+			SourceID:      "279L3gEKqwruBoKGsXZtSVX7vIy",
+			DestinationID: "27CHciD6leAhurSyFAeN4dp14qZ",
+			WorkspaceID:   "279L3V7FSpx43LaNJ0nIs9KRaNC",
+			CreatedAt:     time.Now(),
+		}
+		db := &mockStagingFileSchemaDBRepo{
+			getLatestFunc: func(ctx context.Context, sourceID, destinationID string) (*model.StagingFileSchemaSnapshot, error) {
+				return fresh, nil
+			},
+			insertFunc: func(ctx context.Context, sourceID, destinationID, workspaceID string, schemaBytes json.RawMessage) (uuid.UUID, error) {
+				return fresh.ID, nil
+			},
+		}
+		cache := NewStagingFileSchema(config.New(), db, &fixedExpiryStrategy{expired: true})
+		cache.cache.Put(cacheKey("279L3gEKqwruBoKGsXZtSVX7vIy", "27CHciD6leAhurSyFAeN4dp14qZ"), &model.StagingFileSchemaSnapshot{CreatedAt: time.Now().Add(-time.Hour)}, cache.cacheRefreshTTL())
```

Comment:
`cache.cache.Put` -> Is it possible to use the public `GetOrCreate` method here ?

Reply:
No.

Reply:
If we have Get and Put separate, we can have it. But based on the usecase it's required around the /process API, where we have the modified schema and we want the snapshot shema.